agent/php_agent.h

```diff
@@ -791,8 +793,114 @@ extern bool nr_php_function_is_static_method(const zend_function* func);
+ *           2. The zend_execute_data given by OAPI
+ *
+ * Returns : void
+ *
+ * Note: PHP has a concept of calling files with no function names.  In the
+ *       case of a file being called when there is no function name, the agent
+ *       instruments the file.  In this case, we provide the filename to CLM
+ *       as the "function" name.
+ *       Current CLM functionality only works with PHP 7+
+ */
+extern void nr_php_txn_add_code_level_metrics(
+    nr_attributes_t* attributes,
+    const nr_php_execute_metadata_t* metadata);
+
 #if ZEND_MODULE_API_NO >= ZEND_7_0_X_API_NO /* PHP7+ */
 
+#define NR_ZEND_USER_FUNC_EXISTS(x) \
+  (x && (!x->func || !ZEND_USER_CODE(x->func->type)))
+
+/*
+ * Purpose : Return a pointer to the function name of zend_execute_data.
+ *
+ * Params  : 1. zend_execute_data.
+ *
+ * Returns : A pointer to string, ownership of does NOT pass to the caller and
+ *           string must be dupped if it needs to persist,
+ *           or NULL if the zend_execute_data is invalid.
+ *
+ */
+static inline const char* nr_php_zend_execute_data_function_name(
+    const zend_execute_data* execute_data) {
+  zend_string* function_name = NULL;
+
+  if ((NULL == execute_data) || (NULL == execute_data->func)) {
+    return NULL;
+  }
+
+  function_name = execute_data->func->common.function_name;
+
+  if ((NULL == function_name)
+      && (ZEND_USER_FUNCTION == execute_data->func->type)) {
+    /*
+     * This is the case of a filename being called so there is no function name.
+     * It is broken out separately here in case we need to do something special
+     * with it in the future.
+     */
+    return NULL;
+  }
+  return function_name ? ZSTR_VAL(function_name) : NULL;
+}
+
+/*
+ * Purpose : Return a pointer to the filename of zend_execute_data.
+ *
+ * Params  : 1. zend_execute_data.
+ *
+ * Returns : A pointer to string, ownership of does NOT pass to the caller and
+ *           string must be dupped if it needs to persist,
+ *           or NULL if the zend_execute_data is invalid.
+ *
+ */
+static inline const char* nr_php_zend_execute_data_filename(
+    const zend_execute_data* execute_data) {
+  zend_string* filename = NULL;
+
+  while (NR_ZEND_USER_FUNC_EXISTS(execute_data)) {
+    execute_data = execute_data->prev_execute_data;
+  }
+  if (execute_data) {
+    filename = execute_data->func->op_array.filename;
+  }
+  return filename ? ZSTR_VAL(filename) : NULL;
+}
+
+/*
+ * Purpose : Return a pointer to the scope(i.e., class) name of
+ * zend_execute_data.
+ *
+ * Params  : 1. zend_execute_data.
+ *
+ * Returns : A pointer to string, ownership of does NOT pass to the caller and
+ *           string must be dupped if it needs to persist,
+ *           or NULL if the zend_execute_data is invalid.
+ *
+ */
+static inline const char* nr_php_zend_execute_data_scope_name(
+    const zend_execute_data* execute_data) {
+  zend_class_entry* ce = NULL;
+
+  while (execute_data) {
```

Comment:
How deep can this walk of execute_data get?  Just wondering if it would be possible to walk execute_data and get filename (like in `nr_php_zend_execute_data_filename()`) and scope from a single pass?

Reply:
I was asking if both pieces could be extracted simultaneously from a single loop through the execute data linked list  but it sounds like there wouldn't be much performance benefit.